repos/system_upgrade/el7toel8/actors/selinux/selinuxcontentscanner/libraries/library.py

```diff
@@ -0,0 +1,200 @@
+        run(['sed', '-i', '/%s/s/^/;/g' % r'\|'.join(REMOVED_TYPES_), name])
+        return removed.get("stdout", [])
+    except CalledProcessError:
+        return []
+
+
+def listSELinuxModules():
+    '''
+    Produce list of SELinux policy modules
+
+    Returns list of tuples (name,priority)
+    '''
+    try:
+        semodule = run(['semodule', '-lfull'], split=True)
+    except CalledProcessError:
+        return []
+
+    modules = []
+    for module in semodule.get("stdout", []):
+        # Matching line such as "100 zebra             pp "
+        # "<priority> <module name>    <module type - pp/cil> "
+        m = re.match(r'([0-9]+)\s+([\w-]+)\s+([\w-]+)\s*\Z', module)
+        if not m:
+            # invalid output of "semodule -lfull"
+            api.current_logger().info('Invalid output of "semodule -lfull": %s', module)
+            continue
+        modules.append((m.group(2), m.group(1)))
+
+    return modules
+
+
+def getSELinuxModules():
+    '''
+    Read all custom SELinux policy modules from the system
+
+    Returns 3-tuple (modules, retain_rpms, install_rpms)
+    where "modules" is a list of "SELinuxModule" objects,
+    "retain_rpms" is a list of RPMs that should be retained
+    during the upgrade and "install_rpms" is a list of RPMs
+    that should be installed during the upgrade
+
+    '''
+
+    modules = listSELinuxModules()
+    semodule_list = []
+    # list of rpms containing policy modules to be installed on RHEL 8
+    retain_rpms = []
+    install_rpms = []
+
+    # modules need to be extracted into cil files
+    # cd to /tmp/selinux and save working directory so that we can return there
+
+    # clear working directory
+    rmtree(WORKING_DIRECTORY, ignore_errors=True)
+
+    try:
+        wd = os.getcwd()
+        os.mkdir(WORKING_DIRECTORY)
+        os.chdir(WORKING_DIRECTORY)
+    except OSError:
+        api.current_logger().info("Failed to access working directory! Aborting.")
+        return ([], [], [])
+
+    for (name, priority) in modules:
+        if priority == "200":
+            # Module on priority 200 was installed by an RPM
+            # Request $name-selinux to be installed on RHEL8
+            retain_rpms.append(name + "-selinux")
+            continue
+        if priority == "100":
+            # module from selinux-policy-* package - skipping
+            continue
+        # extract custom module and save it to SELinuxModule object
+        try:
+            run(['semodule', '-c', '-X', priority, '-E', name])
+            # check if the module contains invalid types and remove them if so
+            removed = checkModule(name + ".cil")
+
+            # get content of the module
+            try:
+                with open(name + ".cil", 'r') as cil_file:
+                    module_content = cil_file.read()
+            except OSError as e:
+                api.current_logger().info("Error reading %s.cil : %s", name, e.get("stderr", ""))
```

Comment:
I guess that this will cause traceback as OEError do not have `get` method.

Reply:
Will fix.